### ISSUE TITLE
plot_histogram legend does not show all keys

### ISSUE THREAD

**Issue by Astri-Cornish** [Qiskit/qiskit-terra#10158]
### Environment

- **Qiskit Terra version**: 0.24.0
- **Python version**: 3.9.12
- **Operating system**: Mac OS

### What is happening?

Given I have multiple quasi-distributions and I want to visualise them all at once in a histogram, when I use `plot_histogram`, I expect that the legend will have keys for each of the different quasi-distributions. 

I have this case where one of the quasi-distributions only had counts in 01 and 10, whereas the other distributions had counts across all of the 2-qubit basis states 00, 01, 10, 11 and so because there was no data for 00 for one of the distributions (where 00 is the first bar plotted on the histogram), the legend does not have a key entry for this distribution despite it showing up in the graph for the 10 and 01 states. 

![image](https://github.com/Qiskit/qiskit-terra/assets/65373070/0469535f-f069-403b-a419-74382b628a1e)

I expect a labelled purple box to appear in the legend.

### How can we reproduce the issue?

```python
from qiskit.visualization import plot_histogram
dist_1 = {'01': 0.369, '00': 0.13975, '10': 0.45875, '11': 0.0325}
dist_2 = {'00': 0.011835180722095105,
 '01': 0.48784140401270437,
 '10': 0.4983486948083151,
 '11': 0.001974720456885419}
dist_3 = {'01': 0.4999999999999999, '10': 0.4999999999999999}

legends = ["1", "2", "3"]
plot_histogram([dist_1, dist_2, dist_3], legend=legends)
```

### What should happen?

I expect the same graph with a labelled purple box additionally appearing in the legend for `dist_3` called "3".

**Comment by HuangJunye** [comment]
I can reproduce the issue. This issue seems to appear only when `00` has `0` probabilities.

```python
from qiskit.visualization import plot_histogram
dist_1 = {'00': 0, '01': 1, '10': 1, '11': 1}
dist_2 = {'00': 1, '01': 0, '10': 1, '11': 1}
dist_3 = {'00': 1, '01': 1, '10': 0, '11': 1}
dist_4 = {'00': 1, '01': 1, '10': 1, '11': 0}

legends = ["1", "2", "3", "4"]
plot_histogram([dist_1, dist_2, dist_3, dist_4], legend=legends)
```

![image](https://github.com/Qiskit/qiskit-terra/assets/7631333/955fd358-d390-48ee-98b2-2527a0c7999c)

**Comment by Dpbm** [comment]
I did some tests and I found that the problem is happening in this area of the `_plotting_core` function:

```python
for item, _ in enumerate(data):
        for idx, val in enumerate(all_pvalues[item]):
            label = None
           #every time that we start iterating over all_pvalues[item] and the actual index is 0, we set the label
            if not idx and legend: 
                label = legend[item] 
            if val > 0:
                #However, if the first value is not >0, the legend is not updated, so the loop continues and reset label=None
                rects.append(
                    ax.bar(
                        idx + item * width,
                        val,
                        width,
                        label=label,
                        color=color[item % len(color)],
                        zorder=2,
                    )
                )
```

So, if you change your data to: 

```python
from qiskit.visualization import plot_histogram
dist_1 = {'01': 0.369, '00': 0.13975, '10': 0.45875, '11': 0.0325}
dist_2 = {'00': 0.011835180722095105,
 '01': 0.48784140401270437,
 '10': 0.4983486948083151,
 '11': 0.001974720456885419}
dist_3 = {'01': 0.4999999999999999, '10': 0.4999999999999999, '00':0.000000000001}

legends = ["1", "2", "3"]
plot_histogram([dist_1, dist_2, dist_3], legend=legends)
```
the plot will be:
![result](https://github.com/Qiskit/qiskit-terra/assets/75098594/0e050ab5-e8c9-4f14-9d32-bca502339e11)



But, this is just a simple trick to bypass the problem, the correct way is waiting for a new release.
In my tests, I've changed the `_plotting_core` function to:
```python
for item, _ in enumerate(data):
        label = None

        for idx, val in enumerate(all_pvalues[item]):
            if not label and not idx and legend:
                label = legend[item]
            if val > 0:
                rects.append(
                    ax.bar(
                        idx + item * width,
                        val,
                        width,
                        label=label,
                        color=color[item % len(color)],
                        zorder=2,
                    )
                )
                label = None
```
But I've just tests in your example, maybe it can break in another ways. 
I'll try to check this code in more datasets, and possibly, I'll be creating a PR.

**Comment by Astri-Cornish** [comment]
Thank you @Dpbm! Yes, the dataset I provided was just a test dataset to show the issue with minimal code. The use case I came across this bug was where I have a bell circuit and compare the perfect theoretical circuit results to the results from real hardware, hence the 00 and 11 states additionally popping up in the real device results. As this is learning material I am creating the code for, it would be best if the plotting function works without any adjustments to the datasets.

**Comment by Dpbm** [comment]
Hi @ufranca, there's no problem!
Actually, I've already opened a PR to fix it, I'm just waiting for a code review.
But, if you want, you can check if my modifications really fix the problem and/or there're other problems after my modification. 

[The PR](https://github.com/Qiskit/qiskit-terra/pull/10176)
[my fork (the main branch is the correct)](https://github.com/Dpbm/qiskit-terra)

**Comment by ufranca** [comment]
Ok, great, I'll check it and let you know if I run into any issues.

**Comment by Andir16** [comment]
@Dpbm I see that your PR #10176 got merged, but the issue is still open. Is there still help required for this issue? :)

**Comment by jakelishman** [comment]
Looks like this was fixed by #10176, thanks.
